### PR TITLE
Feature PKCE

### PR DESCRIPTION
**Pull request by onevcat**

### CODE REVIEW

File: LineSDK/LineSDK/Login/PKCE/PKCE.swift [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    /// impractical to guess the value.  It is RECOMMENDED that the output of
+    /// a suitable random number generator be used to create a 32-octet
+    /// sequence. The octet sequence is then base64url-encoded to produce a
+    /// 43-octet URL safe string to use as the code verifier.
+    ///
+    /// Ref: https://tools.ietf.org/html/rfc7636#section-4.1
+    ///
+    static func generateCodeVerifier() -> Data {
+        var bytes = [UInt8](repeating: 0, count: 32)
+        let status = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
```
Comment (onevcat) [inline LineSDK/LineSDK/Login/PKCE/PKCE.swift]: Can we extract this method for generating random data to a helper method (maybe an extension on `Data`), so we can use it for other parts of the login process, such as `state` or `nonce`. Now they are `UUID` and is not that randomly in cryptography

File: LineSDK/LineSDK/Login/PKCE/PKCE.swift [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        if status != errSecSuccess {
```
Comment (onevcat) [inline LineSDK/LineSDK/Login/PKCE/PKCE.swift]: I see a potential risk here. When the `SecRandomCopyBytes` fails, it leads to an all-0 byte as the `codeVerifierData` and the PKCE would be worked around.

File: LineSDK/LineSDK/Login/PKCE/PKCE.swift (hunk 1 above)
Comment (eJamesLin) [reply LineSDK/LineSDK/Login/PKCE/PKCE.swift]: Good idea~ I've extract the `randomData(bytesCount:)` method.

I checked the checking spec of nonce, there's no specified rule for nonce generation.
https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes

But I did find the open id repo had use the same way to generate nonce, to ensure cryptographically random.
https://github.com/openid/AppAuth-iOS/blob/master/Source/OIDAuthorizationRequest.m#L182

Comment (onevcat) [reply LineSDK/LineSDK/Login/PKCE/PKCE.swift]: Yes. I think 32 bytes of random data should be totally fine here!

**Comment by onevcat** [comment]
LGTM! Let's wait for server implementation then we can merge it.

**Comment by onevcat** [comment]
[LGTM](https://www.lgtm.app/i/7AwhYVcRh)

**Pull request merged**